KarhooUISDK/Screens/CheckoutScreen/Checkout/KarhooCheckoutViewModel.swift

```diff
@@ -0,0 +1,419 @@
+                    self?.checkIfNeedsToUpdateState()
+                }
+                .store(in: &cancellables)
+        }
+
+        passangerDetailsViewModel.onTap = { [weak self] in
+            self?.showPassengerDetails()
+        }
+        
+        flightNumberCellViewModel.onTap = { [weak self] in
+            self?.showFlightNumberBottomSheet()
+        }
+
+        flightNumberCellViewModel.flightNumberSubject
+            .dropFirst()
+            .sink{ [weak self] trainNumber in
+                self?.bookingWorker.update(flightNumber: trainNumber)
+            }
+            .store(in: &cancellables)
+
+        trainNumberCellViewModel.onTap = { [weak self] in
+            self?.showTrainNumberBottomSheet()
+        }
+
+        trainNumberCellViewModel.trainNumberSubject
+            .dropFirst()
+            .sink{ [weak self] trainNumber in
+                self?.bookingWorker.update(trainNumber: trainNumber)
+            }
+            .store(in: &cancellables)
+        
+        commentCellViewModel.onTap = { [weak self] in
+            self?.showCommentBottomSheet()
+        }
+
+        commentCellViewModel.commentSubject
+            .dropFirst()
+            .sink{ [weak self] comment in
+                self?.bookingWorker.update(comment: comment)
+            }
+            .store(in: &cancellables)
+    }
+
+    private func setupInitialState() {
+        state = .gatheringInfo
+    }
+
+    private func handleStateUpdate(_ state: CheckoutState) {
+        showError = false
+        switch state {
+        case .loading:
+            // In this stage the view is showing loading overlay
+            break
+        case .gatheringInfo:
+            confirmButtonTitle = UITexts.Booking.next.uppercased()
+        case .readyToBook:
+            confirmButtonTitle = UITexts.Booking.pay.uppercased()
+        case .error(title: let title, message: let message):
+            showError = true
+            errorToPresent = (title: title, message: message)
+        }
+    }
+
+    private func submitBooking() {
+        withAnimation {
+            state = .loading
+        }
+        bookingWorker.performBooking()
+    }
+
+    private func handleBookingState(_ bookingState: CheckoutBookingState) {
+        switch bookingState {
+        case .idle:
+            withAnimation {
+                state = validateIfAllRequiredDataAreProvided() ? .readyToBook : .gatheringInfo
+            }
+        case .loading:
+            break
+        case .failure(let error):
+            state = .error(title: UITexts.Generic.error, message: error.localizedMessage)
+        case .success(let tripInfo):
+            quoteValidityWorker.invalidate()
+            router.routeSuccessScene(
+                with: tripInfo,
+                journeyDetails: journeyDetails,
+                quote: quote,
+                loyaltyInfo: .init(shouldShowLoyalty: false, loyaltyPoints: 0, loyaltyMode: .none)
```

Comment:
We shouldn't be hardcoding values here. We're losing loyalty info.

Reply:
Added a method to LoyaltyServce:
```
func getBasicLoyaltyInfo() -> KarhooBasicLoyaltyInfo {
        let loyaltyModel = modelSubject.value.getSuccessValue()
        var points: Int {
            switch modeSubject.value {
            case .earn: return loyaltyModel??.earnAmount ?? 0
            case .burn: return loyaltyModel??.burnAmount ?? 0
            default: return 0
            }
        }
        return KarhooBasicLoyaltyInfo(
            shouldShowLoyalty: isLoyaltyEnabled,
            loyaltyPoints: points,
            loyaltyMode: modeSubject.value
        )
    }
```